# viraligner

Pipeline for analysis of tiled amplicon data generated on local workstations with realtime-basecalling (GPU). This pipeline is based on snakemake and uses conda.
Viraligner can be started alongside sequencing in minKnow. Currently, there is only one primer set: Human Monkeypox Virus from protocol https://www.protocols.io/private/974ACB78F83911EC998C0A58A9FEAC02?step=1.1
### Before sequencing is complete
- Validates input samplesheet
- (Coming...) Starts realtime sequencing monitoring 
### After sequencing is complete
viraligner looks for the sequencing summary file generated by minKnow when a sequencing run is complete. Once the file is found, the following is carried out:
- Performs read filtering
- Runs artic pipeline (https://github.com/artic-network/fieldbioinformatics) using a custom set of primers and reference.
- - Pipeline maps and trims reads with amplicon coverage normalisation, calls variants using nanopolish and builds a consensus sequence based on input reference
- Generated a batch report with coverage graphs
## How to install
1. Install conda (if you haven't already):

    ```
    cd ~
    wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
    chmod +x Miniconda3-latest-Linux-x86_64.sh
    ./Miniconda3-latest-Linux-x86_64.sh
    ```
    
2. We recommend to use mamba (inside conda) for installing packages onto the miniconda environment.
   
   Hint: mamba is simply faster than the stock conda dependency solver and installer.

   ```
   conda install mamba -n base -c conda-forge
   ```

3. Clone this repository
    ```
    # install viraligner in the home directory
    cd ~
    
    git clone https://github.com/KMA-Aarhus/viraligner.git 
    ```
4. Install snakemake and other dependencies needed for viraligner:

    We recommend that you install viraligner in its own enclosed environment. This is easiest done using the provided environment.yaml file. The following command call creates a conda environment, and install viraligner into it.

    ```    
    conda create -n viraligner && conda activate viraligner && mamba env update --file ~/viraligner/environment.yaml 
    ```
    
Please create an issue (in this repo) if you encounter any problems or unanswered questions during installation or use.
## How to run
viraligner needs a samplesheet of expected samples, the run directory where minKnow saves the raw data, a primer directory, a primer scheme name and a scheme version. These can be set in config.yaml file or provided via command line. We recommend using bash extensions. For an example, see scripts/bash_extensions.sh start_viraligner_mpx
<p>For example, in short to start viraligner with hMPX-2022 primers:</p>

   ```
    cd ~/viraligner
    conda activate viraligner
    snakemake --profile default --use-conda --conda-frontend mamba --config scheme_directory="primer_schemes" scheme_name="hMPX-2022" scheme_version="1"    samplesheet="samplesheet.tsv" rundir="example_rundir"
   ```
## Affiliations

viraligner is developed at Department of Clinical Microbiology (KMA) at Aarhus Universityhospital, Denmark. Parts of this protocol may be site specific but we aim to collaborate when possible with other microbiology departments. 
